packages/panzer/disc-fe/src/Panzer_BasisValues2_impl.hpp

```diff
@@ -45,1711 +45,1628 @@
 
-    // =============================================
-    // Load Reference Values
+    const std::pair<int,int> cell_range(0,num_evaluate_cells_);
+    auto s_aux = Kokkos::subview(aux.get_view(),cell_range,Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
+    auto s_cw = Kokkos::subview(cubature_weights_.get_view(),cell_range,Kokkos::ALL());
+    auto s_bv = Kokkos::subview(bv.get_view(),cell_range,Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
 
-    intrepid_basis->getValues(cell_basis_ref_scalar.get_view(),cell_cub_points.get_view(),Intrepid2::OPERATOR_VALUE);
+    fst::multiplyMeasure(s_aux, s_cw, s_bv);
 
-    if(compute_derivatives){
-      intrepid_basis->getValues(cell_grad_basis_ref.get_view(),cell_cub_points.get_view(),Intrepid2::OPERATOR_GRAD);
-    }
+    // Store for later if cache is enabled
+    CACHE_DATA(weighted_basis_vector);
 
-    // =============================================
-    // Transform reference values to physical values
+    return aux;
 
-    fst::HGRADtransformVALUE(cell_basis_scalar.get_view(),cell_basis_ref_scalar.get_view());
-    typename PHX::Device().fence();
-    for(int b=0;b<num_basis;++b)
-      for(int p=0;p<num_points;++p)
-        basis_scalar(cell,b,p)=cell_basis_scalar(0,b,p);
+  } else {
 
-    if(compute_derivatives){
-        fst::HGRADtransformGRAD(cell_grad_basis.get_view(),cell_jac_inv.get_view(),cell_grad_basis_ref.get_view());
-        typename PHX::Device().fence();
-        for(int b=0;b<num_basis;++b)
-          for(int p=0;p<num_points;++p)
-            for(int d=0;d<num_dim;++d)
-              grad_basis(cell,b,p,d)=cell_grad_basis(0,b,p,d);
-    }
-    // =============================================
-  }
+    const auto element_space = getElementSpace();
+    TEUCHOS_ASSERT(element_space == PureBasis::HCURL || element_space == PureBasis::HDIV);
+    TEUCHOS_ASSERT(num_dim != 1);
 
-  if(build_weighted){
-    const std::pair<int,int> cell_range(0,num_cells);
-    auto s_weighted_basis_scalar = Kokkos::subview(weighted_basis_scalar.get_view(),cell_range,Kokkos::ALL(),Kokkos::ALL());
-    auto s_weighted_measure = Kokkos::subview(weighted_measure.get_view(),cell_range,Kokkos::ALL());
-    auto s_basis_scalar = Kokkos::subview(basis_scalar.get_view(),cell_range,Kokkos::ALL(),Kokkos::ALL());
-    fst::multiplyMeasure(s_weighted_basis_scalar,s_weighted_measure,s_basis_scalar);
-    if(compute_derivatives){
-      auto s_weighted_grad_basis = Kokkos::subview(weighted_grad_basis.get_view(),cell_range,Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
-      auto s_grad_basis = Kokkos::subview(grad_basis.get_view(),cell_range,Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
-      fst::multiplyMeasure(s_weighted_grad_basis,s_weighted_measure,s_grad_basis);
+    // HDIV and HCURL have unique jacobian requirements
+    if(element_space == PureBasis::HCURL){
+      TEUCHOS_ASSERT(cubature_jacobian_inverse_.size() > 0);
```

Comment:
I'm a little worried about the size() > 0 checks. In the past with Tpetra and Epetra, we have had cases where element blocks (or boundary conditions) don't exist on certain MPI processes. this results in having zero indices on some mpi processes. Rather than add complex logic to check that at least one element exists we allow the code to run through construction with zero elements. I'm not sure we are in the same case here, but we might want to allow a workset of zero elements to be created so that we don't need checks all over the code for the number of elements. That would require using "is_evaluated" flags instead of checking that the size is greater than zero. I can envision this passing on our unit tests but failing on larger more complex problems.

Reply:
Yup, good call. We had to use the "is_evaluated" bool for the lazy evaluation stuff in empire for a similar reason. It will also make some of the above comments easier to deal with. I was hoping to avoid adding a bunch of bools though :(

Reply:
This is complete. I added a reset call that you may want to check out. I'm unsure how to de-allocate the MDField objects (I wasn't aware how they wrap Kokkos), but we can revisit that in the future.